### DISCUSSION TITLE
How to avoid dynamic library generated through soft link in directory _solib_xx with directory prefix

### DISCUSSION THREAD

**Discussion by Distancess** [bazelbuild/bazel#18765] [Q&A]
I trying to build a compile a dynamic library(calls A) and it depends on another dynamic library(calls B).
when i finish compile, i use `ldd` to check the dependencies of A and i found that the dependency name of B had an extra diretory prefix, such as `libexternal_SB_SlibB.so` while in external/B, the dynamic library name is libB.so.
is there a way to avoid this problem through the cc_library or cc_shared_library args.

[the remark this answers is not in the document]
**Reply by Distancess** [reply]
can you provide more infomation?

[the remark this answers is not in the document]
**Reply by oquenchil** [reply]
Provide a BUILD file example showing the problem please.

[the remark this answers is not in the document]
**Reply by Distancess** [reply]
This is a simple example
bazel version: 6.1.0
os: centos
the directory struct:
![image](https://github.com/bazelbuild/bazel/assets/16832192/db35a825-49f0-4e10-ab0e-58bfc8b07d2d)
in the root BUILD FILE，the content is：
```
cc_shared_library(
    name = "A",
    roots = [":A_so"],
    dynamic_deps = ["//B:B"]
)

cc_library(
    name = "A_so",
    srcs = ["a.c"],
    deps = ["//B:B_so"],
)
```
in the directory B, the BUILD file is:
```
package(
    default_visibility = [
        "//visibility:public",
    ],
)

cc_shared_library(
    name = "B",
    roots = [":B_so"],
)

cc_library(
    name = "B_so",
    srcs = ["b.c"],
    hdrs = ["b.h"],
    includes = ["B"],
    defines = ["MARCO_DEFINE"],
)
```
and build cmd is `bazel  build --experimental_cc_shared_library //:A`
my question is in the output directory `bazel-bin/B` the dynamic library name is `libB.so` which is expected, but in the directory `_solib_k8` the directory prex is added to the library, which is call `libB_SlibB.so`. I expect it to be `libB.so` because if you use ldd to find the deps of `libA.so`, you can find it links to `libB_SlibB.so` not `libB.so`.

[the remark this answers is not in the document]
**Reply by oquenchil** [reply]
I filed #18790. Fix is on the way.

Thanks for reporting this.

[the remark this answers is not in the document]
**Reply by Distancess** [reply]
thanks for fixed, i'll try it later. by the way, it seem that the `soname` attribute of the dynamic library is not added to `cc_shared_library`. in the preceding example, the soname attribute can't be found through readelf cmd.
![image](https://github.com/bazelbuild/bazel/assets/16832192/fa05a14e-bb89-43f3-9be7-c4afcdf46b07)
here is a normal example compile by makefile
![image](https://github.com/bazelbuild/bazel/assets/16832192/a7b48e9a-0ba2-45ba-9a3f-c9f8f28c45e7)

[the remark this answers is not in the document]
**Reply by oquenchil** [reply]
I filed: https://github.com/bazelbuild/bazel/issues/18798

This is an incompatible change. See the work around in #18798 using `cc_shared_library.user_link_flags`. If you have other issues related to cc_shared_library but unrelated to `soname` please feel free to continue pointing them out here in this thread. If it's related to `soname`, please continue the discussion in the issue.
